### ISSUE TITLE
sky launch error when path contains space

### ISSUE THREAD

**Issue by lhqing** [skypilot-org/skypilot#1246]
Hi Skypilot team,

I'm using google drive to sync `.sky` dir and soft link to '~/.sky' (for #1211 ), it was fine, but today I got this `subprocess.CalledProcessError` error that seems to be due to my google drive path having a space in it.

Thanks!

```
# hanqing @ hanqings-MacBook-Pro in ~/src/whole_mouse_brain/gcp/sky on git:main o [19:21:02] C:1
$ sky launch -c jupyter -y sky-jupyter.yaml
Task from YAML spec: sky-jupyter.yaml
Running task on cluster jupyter...
I 10-13 19:21:33 cloud_vm_ray_backend.py:903] To view detailed progress: tail -n100 -f /Users/hanqing/sky_logs/sky-2022-10-13-19-21-33-522156/provision.log
I 10-13 19:21:33 cloud_vm_ray_backend.py:744] Cluster 'jupyter' (status: STOPPED) was previously launched in GCP (us-west1). Relaunching in that region.
I 10-13 19:21:35 authentication.py:240] OS Login is enabled for GCP project prod-635e. Running additional authentication steps.
cp: /Users/hanqing/My: No such file or directory
cp: Drive/work/sky_status/generated/jupyter.yml: No such file or directory
NAME       LAUNCHED   RESOURCES                STATUS   AUTOSTOP  COMMAND
l-jupyter  3 hrs ago  1x GCP(n2d-standard-96)  STOPPED  -         sky launch -c l-jupyter -y...
jupyter    4 hrs ago  1x GCP(n2d-standard-32)  STOPPED  -         sky launch -c jupyter -y ...

Traceback (most recent call last):
  File "/Users/hanqing/miniconda3/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 179, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/cli.py", line 916, in invoke
    return super().invoke(ctx)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 200, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/cli.py", line 1078, in launch
    _launch_with_confirm(
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/cli.py", line 649, in _launch_with_confirm
    sky.launch(
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 200, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 200, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/execution.py", line 290, in launch
    _execute(
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/execution.py", line 191, in _execute
    handle = backend.provision(task,
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 200, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 179, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/backends/backend.py", line 49, in provision
    return self._provision(task, to_provision, dryrun, stream_logs,
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/backends/cloud_vm_ray_backend.py", line 1705, in _provision
    config_dict = provisioner.provision_with_retries(
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 200, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/backends/cloud_vm_ray_backend.py", line 1377, in provision_with_retries
    config_dict = self._retry_region_zones(
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/backends/cloud_vm_ray_backend.py", line 918, in _retry_region_zones
    config_dict = backend_utils.write_cluster_config(
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 200, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/backends/backend_utils.py", line 805, in write_cluster_config
    _optimize_file_mounts(yaml_path)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/backends/backend_utils.py", line 248, in _optimize_file_mounts
    subprocess.run(f'cp -r {all_local_sources} {local_runtime_files_dir}/',
  File "/Users/hanqing/miniconda3/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'cp -r /Users/hanqing/My Drive/work/sky_status/generated/jupyter.yml /var/folders/39/k4llk3392zq9w1pxwq49jgnh0000gn/T/e8b9b9920e678eb57fe146c90ecfebec ~/.config/gcloud/credentials.db ~/.config/gcloud/application_default_credentials.json ~/.config/gcloud/access_tokens.db ~/.config/gcloud/configurations ~/.config/gcloud/legacy_credentials ~/.sky/.sky_gcp_config_default /var/folders/39/k4llk3392zq9w1pxwq49jgnh0000gn/T/tmp9e4380ak/' returned non-zero exit status 1.
(base)

```

This is the error info:
```
cp: /Users/hanqing/My: No such file or directory
cp: Drive/work/sky_status/generated/jupyter.yml: No such file or directory
```

**Comment by concretevitamin** [comment]
Hey @lhqing - what's the git commit of the skypilot repo? @infwinston just merged a fix for a related problem.

**Comment by lhqing** [comment]
I just did git pull and pip install, the error still exists.

```
# this is the git commit
$ git rev-parse HEAD
de1d4d9aa8fc71bc1b530a80c57733abfe8cfda3
```

**Comment by infwinston** [comment]
From the traceback it looks like a different problem than https://github.com/skypilot-org/skypilot/issues/1189.
```
File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/backends/backend_utils.py", line 248, in _optimize_file_mounts
    subprocess.run(f'cp -r {all_local_sources} {local_runtime_files_dir}/',
```
It seems to be the YAML file path itself containing the space. The fix just merged only applies to the filemount section.
I can look into how to fix this.

**Comment by infwinston** [comment]
Just to confirm what went wrong. @lhqing does the error happen after you move your YAML file under Google Drive?

**Comment by lhqing** [comment]
Thanks for the quick reply! I fixed it (just for now) by copying my `.sky` from google drive into `~/.sky`

I think it is because my google drive path is `/Users/hanqing/My Drive/...`
This is the subprocess call that parsed the path incorrectly:
https://github.com/skypilot-org/skypilot/blob/master/sky/backends/backend_utils.py#L248-L250

**Comment by lhqing** [comment]
> Just to confirm what went wrong. @lhqing does the error happen after you move your YAML file under Google Drive which contains a space in the path?

Yes

**Comment by infwinston** [comment]
Thanks for reporting and confirming this issue! I'm looking into how to fix this.

**Comment by infwinston** [comment]
@lhqing I just submitted a PR https://github.com/skypilot-org/skypilot/pull/1247. It'd be very helpful if you can confirm those changes fix the issue.

**Comment by infwinston** [comment]
The issue is auto-closed by github. Please feel free to re-open it if the latest commit still doesn't solve it.

**Comment by lhqing** [comment]
> The issue is auto-closed by github after I merged the PR. Please feel free to re-open it if the latest commit still doesn't solve it.

Yes, it works for me. Thanks!

**Comment by lhqing** [comment]
@infwinston Sorry I said this too early, there is another error due to the same problem:

```
Traceback (most recent call last):
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/backends/backend_utils.py", line 1105, in query_head_ip_with_retries
    out = subprocess_utils.run(
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/common_utils.py", line 200, in _record
    return f(*args, **kwargs)
  File "/Users/hanqing/miniconda3/lib/python3.9/site-packages/sky/utils/subprocess_utils.py", line 26, in run
    return subprocess.run(cmd,
  File "/Users/hanqing/miniconda3/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'ray get-head-ip /Users/hanqing/My Drive/work/sky_status/generated/jupyter.yml' returned non-zero exit status 2.


```

**Comment by infwinston** [comment]
Ah sorry. Apparently there are more than one place we have to fix. Reopening this issue.

**Comment by infwinston** [comment]
@lhqing we just merged a PR (https://github.com/skypilot-org/skypilot/pull/1250) to fix this! let us know if the issue still persists. thanks :)

**Comment by lhqing** [comment]
@infwinston, thank you! 
I tried the up-to-date skypilot, and it works all good for me.

**Comment by concretevitamin** [comment]
Fixed by#1250.
